src/main/java/graphql/normalized/ExecutableNormalizedField.java

```diff
@@ -61,20 +65,108 @@ private ExecutableNormalizedField(Builder builder) {
         this.parent = builder.parent;
     }
 
+    @Deprecated // doesn't really work
     public boolean isConditional(GraphQLSchema schema) {
         if (parent == null) {
             return false;
         }
         return objectTypeNames.size() > 1 || unwrapAll(parent.getType(schema)) != getOneObjectType(schema);
     }
 
+    /**
+     * Determines whether this NF needs a fragment to select the field. However, it considers the parent
+     * output type when determining whether it needs a fragment.
+     * <p>
+     * Consider the following schema
+     *
+     * <pre>
+     * interface Animal {
+     *     name: String
+     *     parent: Animal
+     * }
+     * type Cat implements Animal {
+     *     name: String
+     *     parent: Cat
+     * }
+     * type Dog implements Animal {
+     *     name: String
+     *     parent: Dog
+     *     isGoodBoy: Boolean
+     * }
+     * type Query {
+     *     animal: Animal
+     * }
+     * </pre>
+     * <p>
+     * and the following query
+     *
+     * <pre>
+     * {
+     *     animal {
+     *         parent {
+     *             name
+     *         }
+     *     }
+     * }
+     * </pre>
+     * <p>
+     * Then we would get the following normalized operation tree
+     *
+     * <pre>
+     * -Query.animal: Animal
+     * --[Cat, Dog].parent: Cat, Dog
+     * ---[Cat, Dog].name: String
+     * </pre>
+     * <p>
+     * If we simply checked the {@link #parent}'s {@link #getFieldDefinitions(GraphQLSchema)} that would
+     * point us to {@code Cat.parent} and {@code Dog.parent} whose output types would incorrectly answer
+     * our question whether this is conditional?
+     * <p>
+     * We MUST consider that the output type of the {@code parent} field is {@code Animal} and
+     * NOT {@code Cat} or {@code Dog} as their respective impls would say.
+     */
+    public boolean isConditional(GraphQLSchema schema, @Nullable String parentOutputTypeName) {
+        if (parent == null || parentOutputTypeName == null) {
+            return false;
+        }
+
+        GraphQLType parentOutputType = schema.getType(parentOutputTypeName);
+
+        // Is not conditional if field exists in interface type, and we have specified all impls
+        if (parentOutputType instanceof GraphQLInterfaceType) {
+            GraphQLInterfaceType parentOutputTypeAsInterface = (GraphQLInterfaceType) parentOutputType;
+            List<GraphQLObjectType> interfaceImpls = schema.getImplementations(parentOutputTypeAsInterface);
+            if (interfaceImpls.size() == objectTypeNames.size()) {
+                return parentOutputTypeAsInterface.getField(this.fieldName) == null
```

Comment:
why do we need check for field name existing in the paren type? If it doesn't exist it would imply that the schema doesn't match the ENF and I would throw an Exception normally.

Reply:
Nah this is checking whether the field exists in the interface type e.g.

```
type Query {
  pet: Pet
}
interface Pet {
  name: String
}
type Dog implements Pet {
  name: String
  collar: String
}
```

So the a tree can go

```
- [Query].pet: Pet
-- [Dog].name
-- [Dog].collar
```

We _need_ an inline fragment for `collar` because it's not in `Pet` but we don't need one for `name`.

I think checking number of impls and object type names is not enough? Actually now that I think about it… maybe you're right and we don't nee to check. I'll confirm tomorrow.